Runtime/Cache/UrlCache.cs

```diff
@@ -0,0 +1,169 @@
+        /// <returns>Response either from cache or from the server</returns>
+        public async Task<IHttpResponse> Process(
+            IHttpRequest request, 
+            CachePolicy cachePolicy, 
+            Func<Task<IHttpResponse>> continuationAction)
+        {
+            IHttpResponse response = null;
+
+            // check cache
+            if (GetFromCache(request.Uri.ToString(), out IHttpResponse cachedResponse))
+            {
+                response = cachedResponse;
+            }
+
+            // override
+            if (cachePolicy != null && cachePolicy.ForceExpire)
+            {
+                response = null;
+                Invalidate(request.Uri.ToString());
+            }
+
+            // make request
+            if (response == null)
+            {
+                // didn't get cached response -> we should make a request
+                response = await continuationAction.Invoke();
+
+                if (cachePolicy != null && response.HasNoErrors)
+                {
+                    // response should be cached
+                    PutIntoCache(response.RequestUri.ToString(), response, cachePolicy.Expiration);
+                }
+            }
+
+            return response;
+        }
+
+        public bool GetFromCache(string url, out IHttpResponse data)
+        {
+            data = null;
+
+            if (_cachedData.TryGetValue(url, out var dataFromCache))
+            {
+                data = dataFromCache.Data;
+
+                // mark as from cache
+                if (data is ICachedHttpResponse cachedResponse)
+                {
+                    cachedResponse.IsFromCache = true;
+                }
+
+                // check if not expired
+                if (dataFromCache.Expired)
+                {
+                    data = null;
+                    Invalidate(url);
+                }
+            }
+
+            return data != null;
+        }
+
+        public void PutIntoCache(string url, IHttpResponse data, TimeSpan expireIn)
+        {
+            if (data == null)
+            {
+                // log error
+                UnityEngine.Debug.LogError($"{nameof(UrlCache)} -> InvalidOperation. Cannot add null data to cache");
+                return;
+            }
+
+            if (_cachedData.ContainsKey(url))
+            {
+                if (!_cachedData.TryRemove(url, out CacheData<IHttpResponse> removedData))
+                {
+                    // log error
+                    UnityEngine.Debug.LogError($"{nameof(UrlCache)} -> Couldn't remove previous data from cache for:{url}");
+                }
+            }
+
+            // check if cache size is not exceeded
+            // log amount left
+
+            if (!_cachedData.TryAdd(url,
+                new CacheData<IHttpResponse>(
+                    data,
+                    expireIn,
+                    (data as ICachedHttpResponse).CacheContentSize())))
```

Comment:
This line looks a little risky - maybe we could check if data is `ICachedHttpResponse` right after entering the method?

Reply:
added null check. If it couldn't cast to ICachedHttpResponse then it will be just null